File: plugin/openapi/src/main/java/io/gatehill/imposter/plugin/openapi/service/SpecificationServiceImpl.java
package io.gatehill.imposter.plugin.openapi.service;

import com.atlassian.oai.validator.OpenApiInteractionValidator;
import com.atlassian.oai.validator.model.SimpleRequest;
import com.atlassian.oai.validator.report.LevelResolver;
import com.atlassian.oai.validator.report.SimpleValidationReportFormat;
import com.atlassian.oai.validator.report.ValidationReport;
import com.fasterxml.jackson.core.JsonGenerationException;
import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import io.gatehill.imposter.ImposterConfig;
import io.gatehill.imposter.plugin.openapi.config.OpenApiPluginConfig;
import io.gatehill.imposter.plugin.openapi.util.ValidationReportUtil;
import io.gatehill.imposter.util.MapUtil;
import io.swagger.models.Scheme;
import io.swagger.v3.oas.models.Components;
import io.swagger.v3.oas.models.ExternalDocumentation;
import io.swagger.v3.oas.models.OpenAPI;
import io.swagger.v3.oas.models.Paths;
import io.swagger.v3.oas.models.info.Info;
import io.swagger.v3.oas.models.security.SecurityRequirement;
import io.swagger.v3.oas.models.servers.Server;
import io.swagger.v3.oas.models.tags.Tag;
import io.vertx.core.http.HttpServerRequest;
import io.vertx.core.http.HttpServerResponse;
import io.vertx.ext.web.RoutingContext;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ExecutionException;
import java.util.function.Function;
import java.util.stream.Collectors;

import static com.google.common.collect.Lists.newArrayList;
import static com.google.common.collect.Maps.newHashMap;
import static java.util.Objects.isNull;
import static java.util.Objects.nonNull;
import static java.util.Objects.requireNonNull;
import static java.util.Optional.ofNullable;
import static org.apache.commons.io.IOUtils.LINE_SEPARATOR;

/**
 * @author Pete Cornish {@literal <outofcoffee@gmail.com>}
 */
public class SpecificationServiceImpl implements SpecificationService {
    private static final Logger LOGGER = LogManager.getLogger(SpecificationServiceImpl.class);
    private static final String DEFAULT_TITLE = "Imposter Mock APIs";
    private static final String ARG_BASEPATH = "openapi.basepath";
    private static final String ARG_SCHEME = "openapi.scheme";
    private static final String ARG_TITLE = "openapi.title";

    private final Cache<String, Object> cache = CacheBuilder.newBuilder().build();
    private final SimpleValidationReportFormat reportFormater = SimpleValidationReportFormat.getInstance();

    @Override
    public OpenAPI getCombinedSpec(ImposterConfig imposterConfig, List<OpenAPI> allSpecs) throws ExecutionException {
        return (OpenAPI) cache.get("combinedSpecObject", () -> {
            try {
                final Scheme scheme = Scheme.forValue(imposterConfig.getPluginArgs().get(ARG_SCHEME));
                final String basePath = imposterConfig.getPluginArgs().get(ARG_BASEPATH);
                final String title = imposterConfig.getPluginArgs().get(ARG_TITLE);

                return combineSpecifications(allSpecs, basePath, scheme, title);

            } catch (Exception e) {
                throw new ExecutionException(e);
            }
        });
    }

    @Override
    public String getCombinedSpecSerialised(ImposterConfig imposterConfig, List<OpenAPI> allSpecs) throws ExecutionException {
        return (String) cache.get("combinedSpecSerialised", () -> {
            try {
                return MapUtil.JSON_MAPPER.writeValueAsString(getCombinedSpec(imposterConfig, allSpecs));
            } catch (JsonGenerationException e) {
                throw new ExecutionException(e);
            }
        });
    }

    @Override
    public OpenAPI combineSpecifications(List<OpenAPI> specs, String basePath, Scheme scheme, String title) {
        requireNonNull(specs, "Input specifications must not be null");
        LOGGER.debug("Generating combined specification from {} inputs", specs.size());

        final OpenAPI combined = new OpenAPI();

        final Info info = new Info();
        combined.setInfo(info);

        final List<Server> servers = newArrayList();
        combined.setServers(servers);

        final List<SecurityRequirement> security = newArrayList();
        combined.setSecurity(security);

        final Paths paths = new Paths();
        combined.setPaths(paths);

        final List<Tag> tags = newArrayList();
        combined.setTags(tags);

        final List<ExternalDocumentation> allExternalDocs = newArrayList();
        final List<Components> allComponents = newArrayList();

        final StringBuilder description = new StringBuilder()
                .append("This specification includes the following APIs:");

        specs.forEach(spec -> {
            ofNullable(spec.getInfo()).ifPresent(specInfo -> description
                    .append("\n* **")
                    .append(specInfo.getTitle())
                    .append("**")
                    .append(ofNullable(specInfo.getDescription()).map(specDesc -> " - " + specDesc).orElse("")));

            servers.addAll(getOrEmpty(spec.getServers()));
            security.addAll(getOrEmpty(spec.getSecurity()));
            tags.addAll(getOrEmpty(spec.getTags()));

            if (nonNull(spec.getExternalDocs())) {
                allExternalDocs.add(spec.getExternalDocs());
            }

            if (nonNull(spec.getComponents())) {
                allComponents.add(spec.getComponents());
            }

            paths.putAll(getOrEmpty(spec.getPaths()));
        });

        // info
        info.setTitle(ofNullable(title).orElse(DEFAULT_TITLE));
        info.setDescription(description.toString());

        // external docs
        final ExternalDocumentation externalDocs = new ExternalDocumentation();

        externalDocs.setDescription(allExternalDocs.stream()
                .map(externalDocumentation -> ofNullable(externalDocumentation.getDescription()).orElse(""))
                .collect(Collectors.joining(LINE_SEPARATOR)));

        // NOTE: The OAS spec only permits a single URL, so, to avoid confusion,
        // we don't set it at all.

        combined.setExternalDocs(externalDocs);

        // components
        final Components components = new Components();

        components.setCallbacks(aggregate(allComponents, Components::getCallbacks));
        components.setExamples(aggregate(allComponents, Components::getExamples));
        components.setExtensions(aggregate(allComponents, Components::getExtensions));
        components.setHeaders(aggregate(allComponents, Components::getHeaders));
        components.setLinks(aggregate(allComponents, Components::getLinks));
        components.setParameters(aggregate(allComponents, Components::getParameters));
        components.setRequestBodies(aggregate(allComponents, Components::getRequestBodies));
        components.setResponses(aggregate(allComponents, Components::getResponses));
        components.setSchemas(aggregate(allComponents, Components::getSchemas));
        components.setSecuritySchemes(aggregate(allComponents, Components::getSecuritySchemes));

        combined.setComponents(components);

        setServers(combined, servers, scheme, basePath);

        combined.setPaths(paths);
        combined.setSecurity(security);
        combined.setTags(tags);

        return combined;
    }

    @Override
    public boolean isValidRequest(
            ImposterConfig imposterConfig,
            OpenApiPluginConfig pluginConfig,
            RoutingContext routingContext,
            List<OpenAPI> allSpecs
    ) {
        if (isNull(pluginConfig.getValidation())) {
            LOGGER.trace("Validation is disabled");
            return true;
        }
        if (!Boolean.TRUE.equals(pluginConfig.getValidation().getRequest())) {
            LOGGER.trace("Request validation is disabled");
            return true;
        }
        if (Boolean.TRUE.equals(pluginConfig.getValidation().getResponse())) {
            throw new UnsupportedOperationException("Response validation is not supported");
        }

        final OpenApiInteractionValidator validator;
        try {
            validator = getValidator(imposterConfig, pluginConfig, allSpecs);
        } catch (ExecutionException e) {
            routingContext.fail(new RuntimeException("Error building spec validator", e));
            return false;
        }

        final HttpServerRequest request = routingContext.request();
        final SimpleRequest.Builder requestBuilder = new SimpleRequest.Builder(request.method().toString(), request.path())
                .withBody(routingContext.getBodyAsString());

        request.headers().forEach(h -> requestBuilder.withHeader(h.getKey(), h.getValue()));

        final ValidationReport report = validator.validateRequest(requestBuilder.build());
        if (!report.getMessages().isEmpty()) {
            final String reportMessages = reportFormater.apply(report);
            LOGGER.warn("Validation failed for {} {}: {}", request.method(), request.path(), reportMessages);

            // only respond with 400 if validation failures are at error level
            if (report.hasErrors()) {
                final HttpServerResponse response = routingContext.response();
                response.setStatusCode(400);

                if (pluginConfig.getValidation().getReturnErrorsInResponse()) {
                    ValidationReportUtil.sendValidationReport(routingContext, reportMessages);
                } else {
                    response.end();
                }
                return false;
            }
        } else {
            LOGGER.debug("Validation passed for {} {}", request.method(), request.path());
        }

        return true;
    }

    /**
     * Returns the specification validator from cache, creating it first on cache miss.
     */
    private OpenApiInteractionValidator getValidator(ImposterConfig imposterConfig, OpenApiPluginConfig pluginConfig, List<OpenAPI> allSpecs) throws ExecutionException {
        return (OpenApiInteractionValidator) cache.get("specValidator", () -> {
            final OpenAPI combined = getCombinedSpec(imposterConfig, allSpecs);

            final OpenApiInteractionValidator.Builder builder = OpenApiInteractionValidator.createFor(combined);

            // custom validation levels
            if (nonNull(pluginConfig.getValidation().getLevels())) {
                LOGGER.trace("Using custom validation levels: {}", pluginConfig.getValidation().getLevels());
                final LevelResolver.Builder levelBuilder = LevelResolver.create();
                pluginConfig.getValidation().getLevels().forEach((key, value) -> levelBuilder.withLevel(key, ValidationReport.Level.valueOf(value)));
                builder.withLevelResolver(levelBuilder.build());
            }

            return builder.build();
        });
    }

    private void setServers(OpenAPI combined, List<Server> servers, Scheme scheme, String basePath) {
        if (nonNull(scheme)) {
            servers.forEach(server -> overrideScheme(scheme.toValue(), server));
        }
        if (nonNull(basePath)) {
            if (servers.isEmpty()) {
                final Server server = new Server();
                servers.add(server);
            }
            servers.forEach(server -> prefixPath(basePath, server));
        }
        combined.setServers(servers.stream().distinct().collect(Collectors.toList()));
    }

    /**
     * Override the scheme of the {@link Server} URL.
     *
     * @param requiredScheme the scheme to set
     * @param server         the server to modify
     */
    private void overrideScheme(String requiredScheme, Server server) {
        try {
            final URI original = new URI(server.getUrl());
            if (nonNull(original.getScheme()) && !original.getScheme().equalsIgnoreCase(requiredScheme)) {
                final URI modified = new URI(
                        requiredScheme,
                        original.getUserInfo(),
                        original.getHost(),
                        original.getPort(),
                        original.getPath(),
                        original.getQuery(),
                        original.getFragment()
                );
                server.setUrl(modified.toASCIIString());
            }
        } catch (URISyntaxException ignored) {
        } catch (Exception e) {
            LOGGER.warn("Error overriding scheme to '{}' for server URL: {}", requiredScheme, server.getUrl());
        }
    }

    /**
     * Prefix the path of the {@link Server} URL.
     *
     * @param basePath the path prefix
     * @param server   the server to modify
     */
    private void prefixPath(String basePath, Server server) {
        try {
            final URI original = new URI(server.getUrl());
            final URI modified = new URI(
                    original.getScheme(),
                    original.getUserInfo(),
                    original.getHost(),
                    original.getPort(),
                    basePath + ofNullable(original.getPath()).orElse(""),
                    original.getQuery(),
                    original.getFragment()
            );
            server.setUrl(modified.toASCIIString());
        } catch (URISyntaxException ignored) {
        } catch (Exception e) {
            LOGGER.warn("Error prefixing scheme with '{}' for server URL: {}", basePath, server.getUrl());
        }
    }

    @SuppressWarnings("unchecked")
    private <T> List<T> getOrEmpty(List<T> list) {
        return ofNullable(list).orElse(Collections.EMPTY_LIST);
    }

    @SuppressWarnings("unchecked")
    private <K, V> Map<K, V> getOrEmpty(Map<K, V> list) {
        return ofNullable(list).orElse(Collections.EMPTY_MAP);
    }

    private <H, T> Map<String, T> aggregate(List<H> allHolders, Function<H, Map<String, T>> mapSupplier) {
        final Map<String, T> all = newHashMap();

        allHolders.stream()
                .map(mapSupplier)
                .filter(Objects::nonNull)
                .forEach(all::putAll);

        return all;
    }
}

File: plugin/openapi/src/main/java/io/gatehill/imposter/plugin/openapi/service/SpecificationServiceImpl.kt
/*
 * Copyright (c) 2016-2021.
 *
 * This file is part of Imposter.
 *
 * "Commons Clause" License Condition v1.0
 *
 * The Software is provided to you by the Licensor under the License, as
 * defined below, subject to the following condition.
 *
 * Without limiting other conditions in the License, the grant of rights
 * under the License will not include, and the License does not grant to
 * you, the right to Sell the Software.
 *
 * For purposes of the foregoing, "Sell" means practicing any or all of
 * the rights granted to you under the License to provide to third parties,
 * for a fee or other consideration (including without limitation fees for
 * hosting or consulting/support services related to the Software), a
 * product or service whose value derives, entirely or substantially, from
 * the functionality of the Software. Any license notice or attribution
 * required by the License must also include this Commons Clause License
 * Condition notice.
 *
 * Software: Imposter
 *
 * License: GNU Lesser General Public License version 3
 *
 * Licensor: Peter Cornish
 *
 * Imposter is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Imposter is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Imposter.  If not, see <https://www.gnu.org/licenses/>.
 */
package io.gatehill.imposter.plugin.openapi.service

import com.atlassian.oai.validator.OpenApiInteractionValidator
import com.atlassian.oai.validator.model.SimpleRequest
import com.atlassian.oai.validator.report.LevelResolver
import com.atlassian.oai.validator.report.SimpleValidationReportFormat
import com.atlassian.oai.validator.report.ValidationReport
import com.fasterxml.jackson.core.JsonGenerationException
import com.google.common.cache.CacheBuilder
import io.gatehill.imposter.ImposterConfig
import io.gatehill.imposter.http.HttpExchange
import io.gatehill.imposter.plugin.openapi.config.OpenApiPluginConfig
import io.gatehill.imposter.plugin.openapi.config.OpenApiPluginValidationConfig.ValidationIssueBehaviour
import io.gatehill.imposter.plugin.openapi.util.ValidationReportUtil
import io.swagger.models.Scheme
import io.swagger.v3.core.util.Json
import io.swagger.v3.oas.models.Components
import io.swagger.v3.oas.models.ExternalDocumentation
import io.swagger.v3.oas.models.OpenAPI
import io.swagger.v3.oas.models.Paths
import io.swagger.v3.oas.models.info.Info
import io.swagger.v3.oas.models.security.SecurityRequirement
import io.swagger.v3.oas.models.servers.Server
import io.swagger.v3.oas.models.tags.Tag
import org.apache.logging.log4j.LogManager
import java.net.URI
import java.net.URISyntaxException
import java.util.*
import java.util.concurrent.ExecutionException
import javax.inject.Inject

/**
 * @author Pete Cornish
 */
class SpecificationServiceImpl @Inject constructor(
    private val imposterConfig: ImposterConfig
) : SpecificationService {
    private val cache = CacheBuilder.newBuilder().build<String, Any>()
    private val reportFormatter = SimpleValidationReportFormat.getInstance()

    @Throws(ExecutionException::class)
    override fun getCombinedSpec(allSpecs: List<OpenAPI>, deriveBasePathFromServerEntries: Boolean): OpenAPI {
        return cache.get("combinedSpecObject") {
            try {
                val scheme = Scheme.forValue(imposterConfig.pluginArgs!![ARG_SCHEME])
                val basePath = imposterConfig.pluginArgs!![ARG_BASEPATH]
                val title = imposterConfig.pluginArgs!![ARG_TITLE]
                return@get combineSpecs(allSpecs, basePath, scheme, title, deriveBasePathFromServerEntries)
            } catch (e: Exception) {
                throw ExecutionException(e)
            }
        } as OpenAPI
    }

    @Throws(ExecutionException::class)
    override fun getCombinedSpecSerialised(allSpecs: List<OpenAPI>, deriveBasePathFromServerEntries: Boolean): String {
        return cache.get("combinedSpecSerialised") {
            try {
                // Use the v3 swagger-core serialiser (io.swagger.v3.core.util.Json) to serialise the spec,
                // to benefit from its various mixins that are not present in the io.swagger.util.Json implementation.
                // In particular, these mixins correctly serialise extensions and components, like SecurityScheme,
                // to their formal values, rather than the Java enum/toString() defaults.
                return@get Json.mapper().writeValueAsString(
                    getCombinedSpec(allSpecs, deriveBasePathFromServerEntries)
                )
            } catch (e: JsonGenerationException) {
                throw ExecutionException(e)
            }
        } as String
    }

    override fun combineSpecs(
        specs: List<OpenAPI>,
        basePath: String?,
        scheme: Scheme?,
        title: String?,
        deriveBasePathFromServerEntries: Boolean
    ): OpenAPI {
        Objects.requireNonNull(specs, "Input specifications must not be null")
        LOGGER.debug("Generating combined specification from {} inputs", specs.size)

        val combined = OpenAPI()
        val info = Info()
        combined.info = info

        val servers = mutableListOf<Server>()
        combined.servers = servers

        val security = mutableListOf<SecurityRequirement>()
        combined.security = security

        val paths = Paths()
        combined.paths = paths

        val tags = mutableListOf<Tag>()
        combined.tags = tags

        val allExternalDocs = mutableListOf<ExternalDocumentation>()
        val allComponents = mutableListOf<Components>()
        val description = StringBuilder().append("This specification includes the following APIs:")

        specs.forEach { spec: OpenAPI ->
            spec.info?.let { specInfo: Info ->
                description
                    .append("\n* **")
                    .append(specInfo.title)
                    .append("**")
                    .append(specInfo.description?.let { specDesc -> " - $specDesc" } ?: "")
            }
            spec.servers?.let(servers::addAll)
            spec.security?.let(security::addAll)
            spec.tags?.let(tags::addAll)
            spec.externalDocs?.let(allExternalDocs::add)
            spec.components?.let(allComponents::add)
            spec.paths?.let(paths::putAll)
        }

        // info
        info.title = title ?: DEFAULT_TITLE
        info.description = description.toString()

        // external docs
        val externalDocs = ExternalDocumentation()
        externalDocs.description = allExternalDocs
            .filter { it.description != null }
            .joinToString(System.lineSeparator()) { it.description }

        // NOTE: The OAS spec only permits a single URL, so, to avoid confusion,
        // we don't set it at all.
        combined.externalDocs = externalDocs

        // components
        val components = Components()
        components.callbacks = aggregate(allComponents) { it.callbacks }
        components.examples = aggregate(allComponents) { it.examples }
        components.extensions = aggregate(allComponents) { it.extensions }
        components.headers = aggregate(allComponents) { it.headers }
        components.links = aggregate(allComponents) { it.links }
        components.parameters = aggregate(allComponents) { it.parameters }
        components.requestBodies = aggregate(allComponents) { it.requestBodies }
        components.responses = aggregate(allComponents) { it.responses }
        components.schemas = aggregate(allComponents) { it.schemas }
        components.securitySchemes = aggregate(allComponents) { it.securitySchemes }
        combined.components = components
        combined.servers = buildServerList(deriveBasePathFromServerEntries, servers, scheme, basePath)
        combined.paths = paths
        combined.security = security
        combined.tags = tags
        return combined
    }

    override fun isValidRequest(
        pluginConfig: OpenApiPluginConfig,
        httpExchange: HttpExchange,
        allSpecs: List<OpenAPI>
    ): Boolean {
        if (Objects.isNull(pluginConfig.validation)) {
            LOGGER.trace("Validation is disabled")
            return true
        }
        if (ValidationIssueBehaviour.IGNORE == pluginConfig.validation?.request) {
            LOGGER.trace("Request validation is disabled")
            return true
        }
        if (ValidationIssueBehaviour.IGNORE != pluginConfig.validation?.response) {
            throw UnsupportedOperationException("Response validation is not supported")
        }

        val validator: OpenApiInteractionValidator = try {
            getValidator(pluginConfig, allSpecs)
        } catch (e: ExecutionException) {
            httpExchange.fail(RuntimeException("Error building spec validator", e))
            return false
        }

        val request = httpExchange.request()
        val requestBuilder = SimpleRequest.Builder(request.method().toString(), request.path())
            .withBody(httpExchange.bodyAsString)

        httpExchange.queryParams().forEach { p -> requestBuilder.withQueryParam(p.key, p.value) }
        request.headers().forEach { h -> requestBuilder.withHeader(h.key, h.value) }

        val report = validator.validateRequest(requestBuilder.build())
        if (report.messages.isNotEmpty()) {
            val reportMessages = reportFormatter.apply(report)
            LOGGER.warn("Validation failed for {} {}: {}", request.method(), request.absoluteURI(), reportMessages)

            // only respond with 400 if validation failures are at error level
            if (report.hasErrors() && ValidationIssueBehaviour.FAIL == pluginConfig.validation.request) {
                val response = httpExchange.response()
                response.setStatusCode(400)
                if (pluginConfig.validation.returnErrorsInResponse) {
                    ValidationReportUtil.sendValidationReport(httpExchange, reportMessages)
                } else {
                    response.end()
                }
                return false
            }
        } else {
            LOGGER.debug("Validation passed for {} {}", request.method(), request.absoluteURI())
        }
        return true
    }

    /**
     * Returns the specification validator from cache, creating it first on cache miss.
     */
    @Throws(ExecutionException::class)
    private fun getValidator(pluginConfig: OpenApiPluginConfig, allSpecs: List<OpenAPI>): OpenApiInteractionValidator {
        return cache.get("specValidator") {
            val combined = getCombinedSpec(allSpecs, pluginConfig.isUseServerPathAsBaseUrl)
            val builder = OpenApiInteractionValidator.createFor(combined)

            // custom validation levels
            pluginConfig.validation?.levels?.let { levels ->
                LOGGER.trace("Using custom validation levels: {}", levels)
                val levelBuilder = LevelResolver.create()
                (defaultValidationLevels + levels).forEach { (key, value) ->
                    levelBuilder.withLevel(key, ValidationReport.Level.valueOf(value))
                }
                builder.withLevelResolver(levelBuilder.build())
            }
            builder.build()
        } as OpenApiInteractionValidator
    }

    private fun buildServerList(
        deriveBasePathFromServerEntries: Boolean,
        servers: List<Server>,
        scheme: Scheme?,
        basePath: String?
    ): List<Server> {
        val finalServers = servers.toMutableList()

        scheme?.let {
            finalServers.forEach { server: Server ->
                overrideScheme(scheme.toValue(), server)
            }
        }
        basePath?.let {
            if (finalServers.isEmpty()) {
                finalServers.add(Server())
            }
            finalServers.forEach { server: Server -> prefixPath(basePath, server) }
        }

        // add a base path for the imposter server URL and each server entry's path component
        val mockEndpoints = mutableListOf<Server>()
        finalServers.forEach { server ->
            mockEndpoints += Server().apply {
                this.url = URI.create(
                    imposterConfig.serverUrl + determineBasePathFromServer(deriveBasePathFromServerEntries, server)
                ).toString()
            }
        }
        // prepend mock endpoints to servers list
        for (i in 0 until mockEndpoints.size) {
            finalServers.add(i, mockEndpoints[i])
        }

        return finalServers.distinct()
    }

    /**
     * Combine the non-null maps of each `Components` object into a single map.
     */
    private fun <H> aggregate(
        allHolders: List<Components>,
        mapSupplier: (Components) -> Map<String, H>?
    ): Map<String, H> {
        val all: MutableMap<String, H> = mutableMapOf()
        allHolders.mapNotNull(mapSupplier).forEach { m -> all.putAll(m) }
        return all
    }

    /**
     * Override the scheme of the [Server] URL.
     *
     * @param requiredScheme the scheme to set
     * @param server         the server to modify
     */
    private fun overrideScheme(requiredScheme: String, server: Server) {
        try {
            val original = URI(server.url)
            if (Objects.nonNull(original.scheme) && !original.scheme.equals(requiredScheme, ignoreCase = true)) {
                val modified = URI(
                    requiredScheme,
                    original.userInfo,
                    original.host,
                    original.port,
                    original.path,
                    original.query,
                    original.fragment
                )
                server.url = modified.toASCIIString()
            }
        } catch (ignored: URISyntaxException) {
        } catch (e: Exception) {
            LOGGER.warn("Error overriding scheme to '{}' for server URL: {}", requiredScheme, server.url)
        }
    }

    /**
     * Prefix the path of the [Server] URL.
     *
     * @param basePath the path prefix
     * @param server   the server to modify
     */
    private fun prefixPath(basePath: String, server: Server) {
        try {
            val original = URI(server.url)
            val modified = URI(
                original.scheme,
                original.userInfo,
                original.host,
                original.port,
                basePath + (original.path ?: ""),
                original.query,
                original.fragment
            )
            server.url = modified.toASCIIString()
        } catch (ignored: URISyntaxException) {
        } catch (e: Exception) {
            LOGGER.warn("Error prefixing scheme with '{}' for server URL: {}", basePath, server.url)
        }
    }

    /**
     * Construct the base path from which the example response will be served.
     *
     * This attempts to derive the path from the first `server` entry in the spec, if one is present
     * and `ImposterConfig.isUseServerPathAsBaseUrl == true`.
     *
     * @param deriveBasePathFromServerEntries whether to derive the path from server entries in the spec
     * @param spec   the OpenAPI specification
     * @return the base path
     */
    override fun determineBasePathFromSpec(spec: OpenAPI, deriveBasePathFromServerEntries: Boolean): String {
        spec.servers.firstOrNull()?.let { firstServer ->
            return determineBasePathFromServer(deriveBasePathFromServerEntries, firstServer)
        }
        return ""
    }

    /**
     * Construct the base path from which the example response will be served.
     *
     * This attempts to derive the path from the `server`, if `ImposterConfig.isUseServerPathAsBaseUrl == true`.
     *
     * @param deriveBasePathFromServerEntries whether to derive the path from server entries in the spec
     * @param server   the Server configuration
     * @return the base path
     */
    private fun determineBasePathFromServer(deriveBasePathFromServerEntries: Boolean, server: Server): String {
        if (deriveBasePathFromServerEntries) {
            // Treat the mock server as substitute for 'the' server.
            // Note: OASv2 'basePath' is converted to OASv3 'server' entries.
            val url = server.url ?: ""
            if (url.length > 1) {
                // attempt to parse as URI and extract path
                try {
                    return URI(url).path
                } catch (ignored: URISyntaxException) {
                }
            }
        }
        return ""
    }

    companion object {
        private val LOGGER = LogManager.getLogger(SpecificationServiceImpl::class.java)
        private const val DEFAULT_TITLE = "Imposter Mock APIs"
        private const val ARG_BASEPATH = "openapi.basepath"
        private const val ARG_SCHEME = "openapi.scheme"
        private const val ARG_TITLE = "openapi.title"

        private val defaultValidationLevels = mapOf(
            "validation.request.parameter.query.unexpected" to "IGNORE"
        )
    }
}